spec/integration/helpers.rb

```diff
@@ -0,0 +1,24 @@
+def apply_manifest(manifest_text)
+  # Applies the given manifest in as safe a manner as possible.
+  # Note: This will try (and fail) to create files, etc.
+  Puppet[:code] = manifest_text
+  node = Puppet::Node.indirection.find(Facter.value(:fqdn))
+  catalog = Puppet::Resource::Catalog.indirection.find(
+    node.name, :use_node => node)
+  catalog = catalog.to_ral
```

Comment:
`puppet-catalog-test` just compiles the catalog. We need to actually attempt to apply it in order to test the custom types and providers.

I was perhaps overstating things a bit in my previous comment. The parts that build the catalog come from the `puppet apply` code, and the rest come from inside the `transaction` object that handles the application of the catalog. We can't just use the transaction ourselves, because that doesn't give us a way to disable the `host_config` flag. As far as I can tell, `catalog.to_ral` does a format conversion to turn a "compiled catalog" data structure into a "catalog that can be applied" object and `catalog.finalize` finishes whatever setup is necessary before we can actually apply.

Unfortunately, puppet is made out of the sort of ruby code that makes spaghetti look friendly and tractable. There's a lot of implicit global state and "action at a distance" that's really hard to disentangle without fairly deep knowledge of the whole system. I *think* I'm using public or semi-public APIs through this whole process (except for resetting `@host_config`, which seems to be unavoidable) but it can be hard to tell. :-(